Allow use of the preferred_username OIDC claim.

meson800 opened the pull request:
Previously, Headscale would only use the `email` OIDC claim to set the Headscale user. In certain cases
(self-hosted SSO), it may be useful to instead use the `preferred_username` to set the Headscale username. This also closes the existing issue #938.

This adds a config setting to use this claim instead. The OIDC docs have been updated to include this entry as well. In addition, this adds an Authelia OIDC example to the docs.

I didn't see any existing unit or integration tests for OIDC, so wasn't sure where to add such tests. If I'm wrong, I'm happy to add them!

I did test this against my SSO server, and confirm that the `preferred_username` grant if you set it works as expected.

<!-- Please tick if the following things apply. You… -->

- [x] read the [CONTRIBUTING guidelines](README.md#contributing)
- [x] raised a GitHub issue or discussed it on the projects chat beforehand (one already exists)
- [ ] added unit tests
- [x] added integration tests
- [x] updated documentation if needed
- [ ] updated CHANGELOG.md

<!-- If applicable, please reference the issue using `Fixes #XXX` and add tests to cover your new code. -->

kradalby commented:
I think this looks sensible, the tests for oidc are located here: https://github.com/juanfont/headscale/blob/main/integration/auth_oidc_test.go

Would be good if we could have one.

meson800 commented:
Thanks. It looks like the current integration tests don't explicitly set claims, but the underlying [mockoidc](https://github.com/oauth2-proxy/mockoidc) does let you queue up successive users with potentially different email/preferred_username claims. I'll try my hand at creating an integration test for this.

vbrandl commented:
Were you able to implement a testcase? I'd love to see this feature since it is blocking me from using OIDC for headscale.

meson800 commented:
@vbrandl I'm working on it, but I'm new to Go and this codebase, and still figuring out how to more efficiently run the integration tests locally without just pushing to Github. The current OIDC tests don't actually check the claim responses at all, so I also need to thread those checks in.

On my stuff, I'm running from these docker images if you want to try them out: https://github.com/meson800/headscale/pkgs/container/headscale

vbrandl commented:
I just tested your image and it does work as expected, but I'd prefer to run the official image for my infrastructure. So I'll wait until this gets merged.

jonathanspw commented:
Thank you for this work @meson800

jonathanspw commented:
I did some testing on this and it appears to work great.  I noticed the email is still cited on the "congrats you've registered your machine" page or whatever it is which might cause some confusion.

That should probably also reflect the username if the setting is enabled.

meson800 commented:
I'll look into changing that welcome page template @jonathanspw  to respect the setting as well. On my local machine I have the mockoidc up and running with new mock users. this will be mergable hopefully within a week or so.

meson800 commented:
@kradalby I added integration tests as requested, so this should be good for review. In sum, this PR includes:

1. Added the option to use the username claim as above, and updated the documentation.
2. Updates to the mockoidc runner; now you can give it a series of users in the format `username_claim <email_claim@example.org>` and have the mockoidc report those users.
3. Using that functionality, there are now integration tests that register clients under `test-user <test-email@example.org>` with the different claim options. I do this in a kind of sketchy way, by checking the FQDN string of `TailscaleClient`s (e.g. makes sure they end in `test-user.headscale.net` or `test-email.headscale.net`, depending on the setting. I picked this method of validating since in the existing `TailscaleClient` interface that `HSIC` implements, there isn't a user field and I didn't want to add them.
4. As suggested by @jonathanspw, I also updated the post-login HTML template to say "Authenticated as `username`, you can now close this window", where `username` is filled in from the preferredUsername or email claim depending on the setting.

Even with a full `nix develop` on both actual Linux and WSL on Windows, I was unable to run one of the go formatters as suggested in the README, and couldn't figure out the path munging needed to install the tool on my own. If I'm failing some format or lint check, I can try to fix on my end if it's easier.

kradalby commented on cmd/headscale/cli/mockoidc.go:
```diff
@@ -63,6 +65,29 @@ func mockOIDC() error {
 		accessTTL = newTTL
 	}
+	
+	mockUsers := os.Getenv("MOCKOIDC_USERS")
```
Would be great if this is prefixed with "HEADSCALE_INTEGRATION_"

meson800 replied:
This is (technically) not just useful for integration tests, as this environment variable controls what mockoidc does if you e.g. set the env variable and ran `headscale mockoidc` yourself. I followed the naming of the other variables in the `mockOIDC` function (`MOCKOIDC_CLIENT_ID`, `MOCKOIDC_CLIENT_SECRET`, and so on). I can still prefix this if desired.

kradalby commented on integration/auth_oidc_test.go:
```diff
@@ -36,6 +37,181 @@ type AuthOIDCScenario struct {
+	oidcMap := map[string]string{
+		"HEADSCALE_OIDC_ISSUER":             oidcConfig.Issuer,
+		"HEADSCALE_OIDC_CLIENT_ID":          oidcConfig.ClientID,
+		"CREDENTIALS_DIRECTORY_TEST":        "/tmp",
```
Is this for Headscale or mockoidc? if its headscale, then prefix would be good.

meson800 replied:
Are you referring to `CREDENTIALS_DIRECTORY_TEST`? I copied this environment map from the TestOIDCAuthenticationPingAll test. For whatever historical reason, that test passes in the client secret with a file instead of directly, which is the reason for the directory. I could simplify if needed (I don't see why `CREDENTIALS_DIRECTORY_TEST` is a separate env variable).

kradalby commented:
Otherwise, looks great!

loprima-l commented:
Isn't it better if we add an internal mail domain for this case, i think this would be easier to deal with in the api and al taht stuff :

ex: 

preferred_username : loprima would make an email like loprima@headscale.sso

It would be more accurate with the Tailscale way of managing sso I guess

meson800 commented:
I think that is something that should be up to your OIDC provider. If
you're self hosting, you can return those types of fake/internal email
addresses with the email grant.

Other self hostable software (Nextcloud, Matrix) deals with the possible
non-uniqueness of the preferred username claim by binding usernames to the
OIDC provider, which is something Headscale doesn't currently do. If
Headscale allows multiple OIDC providers in the future that will have ti be
fixed.

On Thu, Apr 27, 2023, 3:14 PM loprima-l ***@***.***> wrote:

> Isn't it better if we add an internal mail domain for this case, i think
> this would be easier to deal with in the api and al taht stuff :
>
> ex:
>
> preferred_username : loprima would make an email like
> ***@***.***
>
> It would be more accurate with the Tailscale way of managing sso I guess
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/juanfont/headscale/pull/1287#issuecomment-1526216390>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOEWD7LKA676I4VSPD4DO3XDLASHANCNFSM6AAAAAAWIJSWKI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

loprima-l commented:
I agree with that, but aren't we there to implement best practices to follow ? I think using multiple ways of storing a user "id" will be a mess that will complicate the interconnection between Headscale and other pieces of software.

My advice is that we should have one way to identify a user, if we already have email based user distinction, wouldn't it be bad to implement other ways that doesn't fit with the Tailscale way ?

meson800 commented:
The default OIDC settings prior to this PR already strips the domain from
the email claim unless you configure it not to. If you e.g. point Google's
OIDC provider at it, you'll get a Headscale user with @gmail.com removed.
There's still only one type of user ID with or without this PR.

I think there is an advantage for allowing the option of using the
preferred_username claim to create users; this is how OIDC supports letting
SSO users/administrators pick stable usernames despite changing a backing
email address. There is a unique source of truth on which user is which,
the OIDC provider.

I've only used Headscale; can you elaborate on what the Tailscale way to do
this is?

On Thu, Apr 27, 2023, 3:31 PM loprima-l ***@***.***> wrote:

> I agree with that, but aren't we there to implement best practices to
> follow ? I think using multiple ways of storing a user "id" will be a mess
> that will complicate the interconnection between Headscale and other pieces
> of software.
>
> My advice is that we should have one way to identify a user, if we already
> have email based user distinction, wouldn't it be bad to implement other
> ways that doesn't fit with the Tailscale way ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/juanfont/headscale/pull/1287#issuecomment-1526253118>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOEWDZVBYP5IX2JMG2UO3TXDLCR5ANCNFSM6AAAAAAWIJSWKI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

loprima-l commented:
As you can see here : ![See here](https://user-images.githubusercontent.com/69201633/234977645-4f4fbc59-d59e-465a-8bce-ca7594fb82f6.png)

Tailscale is using the provider domain, I think that's a better way to do it, maybe we must implement this way also in Headscale, but it would be necessary to use a headscale prefix, so we don't break up the compatibility with native Tascale APP, I don't know how they refer to the login methods.

meson800 commented:
Thanks for the screenshot!

My feeling is that I think this probably belongs with multi-provider
support, not in this PR. I could work on it though in a future PR though.

I would imagine that with multiple providers, the headscale admin could
configure a suffix for each OIDC provider. E.g. Github could get a @github
prefix, self-hosted Authelia could get a @internal.sso prefix, whatever.

On Thu, Apr 27, 2023, 4:03 PM loprima-l ***@***.***> wrote:

> As you can see here : [image: See here]
> <https://user-images.githubusercontent.com/69201633/234977645-4f4fbc59-d59e-465a-8bce-ca7594fb82f6.png>
>
> Tailscale is using the provider domain, I think that's a better way to do
> it, maybe we must implement this way also in Headscale, but it would be
> necessary to use a headscale prefix, so we don't break up the compatibility
> with native Tascale APP, I don't know how they refer to the login methods.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/juanfont/headscale/pull/1287#issuecomment-1526328046>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOEWDZMQ6ZUBNQVGUSQWQLXDLGKXANCNFSM6AAAAAAWIJSWKI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

loprima-l commented:
I think that would be AMAZING <3 

Like it will help a lot in the infinite quest of having a standardized system, I'm really busy to work on it but if you need help let me know I could write the docs if you want, I'm already planing to open a PR to add a gitbook.

meson800 commented:
I merged against main so this can run tests again.

@kradalby do you want me to change the environment variables as in the review above? I named them to be consistent with the surrounding code.

@juanfont do you want me to update/edit this PR in some way/can you review? I see that there are other OIDC PRs that this might conflict with this (#1435), I can incorporate those changes if needed.

leeaash commented:
When will this PR be merged to main branch, that feature is acutely I want.

joachimtingvold commented:
Yes, please. Immutable identifier to be used by headscale would be nice to have, as most IDPs allows users to change their emails, and thus break their headscale setup/login.

meson800 commented:
I can try to update this branch to get it considered again; a lot has changed since I submitted the PR.

jonathanspw commented:
> I can try to update this branch to get it considered again; a lot has changed since I submitted the PR.

Please do.

kradalby commented:
Hi! as part of #1473, we have reorganised a lot of the code. 

To clear PRs that needs to be rebased or redone, we are closing open PRs that will require significant code change to be merged.

In addition, the issue of the PR might in some cases have been fixed, change or no longer relevant, so it would be great if this is considered as well.

Thank you for your contribution! 

If it is still relevant and the PR is reopened, we will aim at getting the changes into the next release after the reorg if accepted.

meson800 commented:
@kradalby I can't reopen the PR myself, but this is ready to be considered again; I rebased and tested [meson800:oidc_username_claim](https://github.com/meson800/headscale/tree/oidc_username_claim) on top of the re-orged main.

# Summary

- [x] read the [CONTRIBUTING guidelines](https://github.com/juanfont/headscale/pull/README.md#contributing)
- [x] raised a GitHub issue or discussed it on the projects chat beforehand (one already exists)
- [ ] added unit tests
- [x] added integration tests
- [x] updated documentation if needed
- [x] updated CHANGELOG.md

This PR adds the ability to use the `preferred_username` claim, closing #938. I added documentation, updated the CHANGELOG.md, and added integration tests; I didn't add a unit test but can if needed.

The short version is that you can now set the `use_username_claim` config setting inside the `oidc` category. If set to true, the `preferred_username` claim is used to set the Headscale username instead of the `email` claim. I updated the way `mockoidc` is called to let integration tests set these claims.

## Integration tests
In particular, the integration tests are implemented by setting the mock OIDC user to `test-user <test-email@example.com>`, then it checks that the FQDN of the user is [`test-user.headscale.net`](https://github.com/meson800/headscale/blob/oidc_username_claim/integration/auth_oidc_test.go#L104) or [`test-email.headscale.net`](https://github.com/meson800/headscale/blob/oidc_username_claim/integration/auth_oidc_test.go#L191).

#### Test failures
Of note, I ran [all of the integration tests](https://github.com/meson800/headscale/pull/1) in a stub PR in my repo; some integration tests appear flaky so I had to rerun them, but one still fails despite reruns, `TestPingAllByIP`. This one also seem to be failing in other PR runs in this repo so I don't think it's related ([example 1](https://github.com/juanfont/headscale/actions/runs/6647003340/job/18061574738))

## Real tests
I generated a Docker image for this PR and tested it out in my tailnet, and it works, correctly using the `preferred_username` claim:
![image](https://github.com/juanfont/headscale/assets/1854223/d82c47aa-f2b7-4ad1-bb96-acfc76ba10e2)

LarssonOliver commented:
Looks awesome! Would really help my auth setup.

Helvio88 commented:
Any reason this wasn't merged? I see it's closed. I think it's a very welcome improvement.

meson800 commented:
I think that one of the concerns is it wasn't generalizable enough. See this comment: https://github.com/juanfont/headscale/issues/938#issuecomment-1952697228
